lib/graphql_metrics/analyzer.rb

```diff
@@ -0,0 +1,193 @@
+module GraphQLMetrics
+  class Analyzer < GraphQL::Analysis::AST::Analyzer
+    FieldTimingCallback = Struct.new(:static_metrics, :callback, keyword_init: true)
+
+    CONTEXT_NAMESPACE = :graphql_metrics_analysis
+    RUNTIME_METRICS_ENABLED = :runtime_metrics_enabled
+
+    QUERY_START_TIME_KEY = :query_start_time
+    QUERY_START_TIME_MONOTONIC_KEY = :query_start_time_monotonic
+    FIELD_TIMING_CALLBACKS_KEY = :field_timing_callbacks
+    ANALYZER_INSTANCE_KEY = :analyzer_instance
+
+    # NOTE: These constants are used both to match tracing keys from graphql-ruby as well as to store in-line and lazy
+    # field resolution timings in context.
+    GRAPHQL_TRACING_FIELD_KEY = 'execute_field'
+    GRAPHQL_TRACING_LAZY_FIELD_KEY = 'execute_field_lazy'
+
+    class << self
+      def current_time
+        Process.clock_gettime(Process::CLOCK_REALTIME)
+      end
+
+      def current_time_monotonic
+        Process.clock_gettime(Process::CLOCK_MONOTONIC)
+      end
+
+      def before_query(query)
+        return if query.context[:skip_graphql_metrics_analysis]
+        return unless query.valid?
+
+        query.context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+          ns[RUNTIME_METRICS_ENABLED] = true
+          ns[FIELD_TIMING_CALLBACKS_KEY] = []
+          ns[QUERY_START_TIME_KEY] = current_time
+          ns[QUERY_START_TIME_MONOTONIC_KEY] = current_time_monotonic
+          ns[GRAPHQL_TRACING_FIELD_KEY] = {}
+          ns[GRAPHQL_TRACING_LAZY_FIELD_KEY] = {}
+        end
+      end
+
+      def after_query(query)
+        return if query.context[:skip_graphql_metrics_analysis]
+        return unless query.valid?
+
+        query.context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+          duration = current_time_monotonic - ns[QUERY_START_TIME_MONOTONIC_KEY]
+          runtime_metrics = { start_time: ns[QUERY_START_TIME_KEY], duration: duration }
+
+          analyzer = ns[ANALYZER_INSTANCE_KEY]
```

Comment:
Hmm good question.

I went this route because I want to support this analyzer being used on its own, without instrumentation (which calls `before_query`). But now that I think of it, I could probably scrap the logic in `initialize` and do it idempotently either in `before_query` (when instrumentation is being used) or in the first analyzer hook method that notices that setup wasn't done (when not used as instrumentation).

Reply:
I can confirm the gem does create new analyzer instances each time query.result is called.

I thought through options to move setup logic currently in before_query and initialize to other places, but decided to largely leave it as is (haven't pushed local commits yet, will notify when I do):

`before_query` has all of the runtime metrics context key-value setup, since we only need this stuff when instrumentation is opted into.

`initialize` seems like the best place to obtain the analyzer instance to store for later use, as well as the instance vars that no instrumentation hooks need access to.